### PR TITLE
refactor(engine): add threadsafe wrappers to legacy plugin dispatches

### PR DESCRIPTION
**Pull request by mcous**
## Overview

Wild stab in the dark at #8625 

## Changelog

- refactor(engine): add threadsafe wrappers to legacy plugin dispatches

## Review requests

- [x] Repro protocols from #8625 no longer hang

## Risk assessment

Low

### CODE REVIEW

File: api/src/opentrons/protocol_engine/plugins.py [hunk 1]
```diff
@@ -35,6 +36,20 @@ def dispatch(self, action: Action) -> None:
         """
         return self._action_dispatcher.dispatch(action)
 
+    @final
+    def dispatch_threadsafe(self, action: Action) -> None:
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/plugins.py:41]: Does `dispatch_threadsafe()` need to be a concern of all plugins, or could it be an implementation detail of `LegacyContextPlugin`?

File: api/src/opentrons/protocol_engine/plugins.py [hunk 2]
```diff
@@ -35,6 +36,20 @@ def dispatch(self, action: Action) -> None:
...
+
+        Child thread must be created with `anyio.to_thread`.
+
+        Arguments:
+            action: A new ProtocolEngine action to send into the pipeline.
+                This action will flow through all plugins, including
+                this one, so be careful to avoid infinite loops. In general,
+                do not dispatch an action your plugin will react to.
+        """
+        return from_thread.run_sync(self._action_dispatcher.dispatch, action)
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/plugins.py:52]: Following up an in-person discussion about whether we need `anyio` for this: no, I don't think we do.

`LegacyContextPlugin` would keep a handle to the main thread's `loop`. When the background protocol thread emits a broker message, `LegacyContextPlugin` would, still from that background thread, call [`loop.run_coroutine_threadsafe()`](https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe). It would then call `.result()` on the returned future to block until after the dispatch completes.

Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/plugins.py:52]: That said, I'm 50/50 on whether we should use `anyio` or raw `asyncio` here. Your call.

* I like that `anyio.to_thread.run_sync()` is [thoughtful about thread cancellation](https://anyio.readthedocs.io/en/stable/api.html?highlight=to_thread#anyio.to_thread.run_sync).
* However, it *is* another layer atop something that's already pretty complicated.

File: api/src/opentrons/protocol_runner/legacy_context_plugin.py [hunk 3]
```diff
@@ -15,26 +15,6 @@
...
-# background thread. But our reactions will modify Protocol Engine state, which may
-# conflict with unrelated Protocol Engine accesses from the main FastAPI thread.
-#
-#
-# Problem 2 -- unsynchronized concurrent access to legacy ProtocolContext state...
-#
-# We currently add and remove subscriptions on the legacy ProtocolContext in direct
-# reaction to receiving play and stop actions from Protocol Engine. To be correct, we
-# need to guarantee that, when we do this, the legacy ProtocolContext will be inactive.
-# It's not clear whether we currently accomplish this.
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_runner/legacy_context_plugin.py:38]: I know we talked about this before and you reasoned that this second problem isn't actually a thing, but I still find the subscription setup/teardown on `PlayAction`/`StopAction` slightly hard to follow.

Not a strong preference, but how would you feel about a `# todo` for giving `LegacyContextPlugin` a single explicit setup point and a single explicit teardown point? For example, maybe its setup point is `.__init__()` and its teardown point is a new `.close()` method. Or maybe `AbstractPlugin` defines a `.setup()` and `.teardown()` that Protocol Engine guarantees will each be called exactly once.

The things I'm hoping this would make evident are:

* Subscriptions are set up exactly once.
    * This happens before the background protocol thread is made active.
* Later, subscriptions are torn down exactly once.
    * This happens after the background protocol thread ceases to be active.

**Review by SyntaxColoring — COMMENTED** [review]
This looks correct to me and is definitely an improvement. Happy to merge with this approach, but, here are some follow-up concerns that need pointing out.

As I understand it, with the fix in this PR, things work like this:

1. We start running the user-written Python script in its own background thread.
2. In that background thread:
	1. The user script calls a Protocol API method like `aspirate()`.
	2. Inside the `aspirate()` call, the Protocol API calls our mapping callback.
	3. Our mapping callback converts data from the Protocol API into a nice new Protocol Engine command model.
	4. Our mapping callback tells the *main thread* to, at its earliest convenience, dispatch an `UpdateCommandAction`, to insert the new command into Protocol Engine state.
	5. Our mapping callback blocks until the main thread finishes doing that, and then returns.
    6. The `aspirate()` call returns, and the user script continues.

A problem with this is that it makes the progress of the protocol depend on the progress of the main FastAPI thread. If one task hogs the main FastAPI thread for a while, the robot will physically stop moving.

Tasks *do* regularly hog the main FastAPI thread, unfortunately. For example:

* https://github.com/Opentrons/opentrons/issues/6951
* Unticketed: `GET`ing a protocol that has a big analysis takes multiple seconds, and appears to block other requests. I suspect the main FastAPI thread is blocked doing a big copy or serialization.

Of course, these are probably bugs at the `robot-server` FastAPI layer. But they're bugs that we could choose to be resilient to, down here. Here's how:

1. We start running the user-written Python script in its own background thread.
2. In that background thread:
	1. The user script calls a Protocol API method like `aspirate()`.
	2. Inside the `aspirate()` call, the Protocol API calls our mapping callback.
	3. Our mapping callback converts data from the Protocol API into a nice new Protocol Engine command model.
	4. **Our mapping callback appends the new command to a thread-safe queue of unbounded length.**
	5. Our mapping callback returns.
    6. The `aspirate()` call returns, and the user script continues.
3. Simultaneously, in the main FastAPI thread:
    1. **A concurrent background task reads from the thread-safe queue and dispatches those commands into the Protocol Engine.**
    2. Something kills the task once the protocol ends.

That way, even if something blocks the main FastAPI thread, the robot will keep trucking along. The run log will fall behind, but it should eventually catch up unless the main FastAPI thread somehow got stuck in a truly infinite loop.

A danger is that this would add complexity and hide deeper problems. Now that I'm thinking about it, hardware communication is going single-threaded async too, right? So if HTTP requests can block each other, and they can also block hardware communication. The problem with the asyncio cooperative multitasking "everyone take turns" model is that it assumes everything in the event loop is well-behaved, which will get increasingly hard as we bring more and more stuff into the same `robot-server` event loop. Ugh.

File: api/src/opentrons/protocol_engine/plugins.py (hunk 2 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/plugins.py:52]: A wrinkle with using `run_coroutine_threadsafe` is that `_action_dispatcher.dispatch` is synchronous, and thus does not return a coroutine

**Comment by mcous** [comment]
This does not appear to resolve our reproduction case, so closing down for the time being

**Comment by mcous** [comment]
Wait, I take that back! I _am_ able to successfully complete a simulation of the repro protocol on this branch. It just takes about ~2 minutes on PE instead of 30 sec on RPC. I think that makes sense, all the Pydantic models being created are likely introducing overhead, which is a problem we can solve separately.

Going to continue with this branch and implement the requested `setup` / `teardown` methods.

I agree that at a certain point, having some sort of cross-thread stream or queue probably makes sense here. For now, I'd like to sit tight (and maybe get a spike ticket in place). Best case scenario is this janky solution is _just_ good enough to allow us to focus on APIv3 without anything exploding, but I don't feel super confident in that given this issue in the first place

File: api/src/opentrons/protocol_engine/plugins.py (hunk 1 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/plugins.py:41]: I think it's a little easier to guarantee correctness if we put in here, but no strong preference

**Review by sanni-t — APPROVED** [review]
The long protocol completes analysis successfully. Took ~15 seconds on a mac, while ~10minutes on the robot.

**Review by SyntaxColoring — APPROVED** [review]
Code changes since my last review look good to me. Thanks!
